File: web-ts/src/merge.ts
import type {MeshRecord} from "./types.js";
export function winner(a:MeshRecord,b:MeshRecord):MeshRecord{
  if(a.record_id!==b.record_id) throw new Error("record id mismatch");
  if(a.revision!==b.revision) return a.revision>b.revision?a:b;
  if(a.source_priority!==b.source_priority) return a.source_priority>b.source_priority?a:b;
  const ta=Date.parse(a.updated_at),tb=Date.parse(b.updated_at); if(ta!==tb)return ta>tb?a:b;
  return JSON.stringify(a.body)>=JSON.stringify(b.body)?a:b;
}
export function mergeRecords(local:Map<string,MeshRecord>,incoming:Iterable<MeshRecord>):Map<string,MeshRecord>{const out=new Map(local);for(const r of incoming){const old=out.get(r.record_id);out.set(r.record_id,old?winner(old,r):r)}return out}
export function activeRecords(records:Iterable<MeshRecord>,now=new Date()):MeshRecord[]{return [...records].filter(r=>!r.expires_at||Date.parse(r.expires_at)>now.getTime()).sort((a,b)=>a.record_id.localeCompare(b.record_id))}

File: mobile-kotlin/src/main/kotlin/org/civicmesh/sync/Merge.kt
package org.civicmesh.sync

import java.time.Instant

interface Revisioned { val recordId: String; val revision: Int; val updatedAt: String }
data class MeshRecord(override val recordId:String, override val revision:Int, override val updatedAt:String, val type:String, val payload:String, val sourcePriority:Int=0):Revisioned
object MergePolicy {
    fun winner(a: MeshRecord,b: MeshRecord):MeshRecord {
        require(a.recordId==b.recordId){"cannot merge different records"}
        if(a.revision!=b.revision)return if(a.revision>b.revision)a else b
        if(a.sourcePriority!=b.sourcePriority)return if(a.sourcePriority>b.sourcePriority)a else b
        val ta=Instant.parse(a.updatedAt);val tb=Instant.parse(b.updatedAt);if(ta!=tb)return if(ta.isAfter(tb))a else b
        return if(a.payload>=b.payload)a else b
    }
    fun merge(local:Map<String,MeshRecord>,incoming:Iterable<MeshRecord>):Map<String,MeshRecord>{val out=local.toMutableMap();for(record in incoming){val old=out[record.recordId];out[record.recordId]=if(old==null)record else winner(old,record)};return out.toMap()}
}
